transport/src/main/java/io/netty/channel/nio/NioEventLoop.java

```diff
@@ -302,11 +309,15 @@ public void run() {
     @Override
     protected void run() {
         for (;;) {
-            boolean oldWakenUp = wakenUp.getAndSet(false);
             try {
-                if (hasTasks()) {
-                    selectNow();
-                } else {
+                boolean hasTasks = hasTasks();
+                int selectNowCount = selectNow();
+                if (backoff.delaySelect(selectNowCount, hasTasks)) {
```

Comment:
I thought about it a bit more and I think we can change the interface to replicate the existing behavior and also allow for extension. This strategy allows us to avoid an extra system call on every loop iteration in the default case and  will also work with EpollEventLoop.

```java
public interface BackOffStrategy {
  interface SelectNowSupplier {
    int selectNow();
  }
  
  // return value: -2 means delay, -1 means select, >= 0 means 
  int calculateStrategy(SelectNowSupplier supplier, boolean hasTasks);
}

public final class DefaultBackoffStrategy implements BackOffStrategy {
  // this class should be a singleton.
  public int calculateStrategy(SelectNowSupplier supplier, boolean hasTasks) {
    return hasTasks ? supplier.selectNow() : -1;
  }
}


// In NioEventLoop:
private final SelectNowSupplier SELECT_NOW_SUPPLIER = new SelectNowSupplier() {
 public  int selectNow() {
   selectNow();
  }
};

..


selectloop for(;;) {
  try {
    switch(backoff.calculateStrategy(SELECT_NOW_SUPPLIER, hasTasks()) {
    case -2:
        continue selectloop;
    case -1:
        select(wakenUp.getAndSet(false));
        if (wakenUp.get()) {
          selector.wakeup();
         }
      // fall through
     default:
       cancelledKeys = 0;
       needsToSelectAgain = false;
        ...
       break;
    }
```